staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

```diff
@@ -4067,6 +4067,194 @@ func TestValidateCustomResourceDefinition(t *testing.T) {
+			resource: &apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{Name: "plural.group.com"},
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					Group:    "group.com",
+					Version:  "version",
+					Versions: singleVersionList,
+					Scope:    apiextensions.NamespaceScoped,
+					Names: apiextensions.CustomResourceDefinitionNames{
+						Plural:   "plural",
+						Singular: "singular",
+						Kind:     "Plural",
+						ListKind: "PluralList",
+					},
+					Validation: &apiextensions.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							Type: "object",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".a.b.c",
+								},
+							},
+							Properties: map[string]apiextensions.JSONSchemaProps{
+								"a": {
+									Type: "object",
+									Properties: map[string]apiextensions.JSONSchemaProps{
+										"b": {
+											Type: "object",
+											Properties: map[string]apiextensions.JSONSchemaProps{
+												"c": {
+													Type: "number",
+												},
+											},
+										},
+									},
+								},
+								"list": {
+									Type: "array",
+									Items: &apiextensions.JSONSchemaPropsOrArray{
+										Schema: &apiextensions.JSONSchemaProps{
+											Type: "object",
+											Properties: map[string]apiextensions.JSONSchemaProps{
+												"a": {
+													Type: "number",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				Status: apiextensions.CustomResourceDefinitionStatus{
+					StoredVersions: []string{"version"},
+				},
+			},
+			errors: []validationMatch{},
+		},
+		{
+			name: "x-kubernetes-validations have invalid fieldPath",
+			resource: &apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{Name: "plural.group.com"},
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					Group:    "group.com",
+					Version:  "version",
+					Versions: singleVersionList,
+					Scope:    apiextensions.NamespaceScoped,
+					Names: apiextensions.CustomResourceDefinitionNames{
+						Plural:   "plural",
+						Singular: "singular",
+						Kind:     "Plural",
+						ListKind: "PluralList",
+					},
+					Validation: &apiextensions.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							Type: "object",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".list[0].b",
+								},
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".list[0.b",
+								},
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".list0].b",
+								},
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".a.c",
+								},
+								{
+									Rule:      "self.a.b.c > 0.0",
+									FieldPath: ".a.b.d",
```

Comment:
Recommend expanding the testing of the validateFieldPath() parser code somewhere. It can be a smaller dedicated unit test specifically for the parser function.  Some candidate tests:

- missing leading `.`,  e.g. `a.c`. This will be a common mistake, a nice error message here will REALLY help users.
- duplicated `.`, e.g. `.a..b`
- malformed array index, e.g. `.a.b[-1]` or `.a.b[1.0]`
- malformed map key `.a.b["this is a " bad key "]`. 
- - valid/invalid escaping of map string keys
- check that allowed (unicode) chars in field names and map keys work as expected
- whitespace, e.g. `.a  .b`, `.a[ 1 ]` (I don't know if the JSON Path spec allows or disallows this, but we should match the spec)

Reply:
More tests have been added